imxrt-hal/src/can/mod.rs

```diff
@@ -0,0 +1,1040 @@
+        }
+        // message available
+        if (ral::read_reg!(ral::can, self.reg, IFLAG1) & (0x00000020 as u32)) != 0 {
+            return None;
+        }
+        Some(())
+    }
+
+    #[inline(always)]
+    fn mailbox_number_to_address(&self, mailbox_number: u8) -> u32 {
+        self.base_address() + 0x80_u32 + (mailbox_number as u32 * 0x10_u32)
+    }
+
+    #[inline(always)]
+    fn mailbox_number_to_rximr_address(&self, mailbox_number: u8) -> u32 {
+        self.base_address() + 0x880_u32 + (mailbox_number as u32 * 0x4_u32)
+    }
+
+    #[inline(always)]
+    fn mailbox_number_to_idflt_tab_address(&self, mailbox_number: u8) -> u32 {
+        self.base_address() + 0xE0_u32 + (mailbox_number as u32 * 0x4_u32)
+    }
+
+    #[inline(always)]
+    fn read_mailbox_code(&mut self, mailbox_number: u8) -> u32 {
+        let mailbox_addr = self.mailbox_number_to_address(mailbox_number);
+        unsafe { core::ptr::read_volatile(mailbox_addr as *const u32) }
+    }
+
+    #[inline(always)]
+    fn read_mailbox(&mut self, mailbox_number: u8) -> Option<MailboxData> {
+        if (self.read_imask() & (1_u64 << mailbox_number)) != 0 {
+            return None;
+        }
+
+        let code = self.read_mailbox_code(mailbox_number);
+        let c = FlexCanMailboxCSCode::from_code_reg(code);
+
+        let mailbox_addr = self.mailbox_number_to_address(mailbox_number);
+
+        match c {
+            // return None from a transmit mailbox
+            c if c.is_tx_mailbox() => None,
+            // full or overrun
+            c if (c == FlexCanMailboxCSCode::RxFull) | (c == FlexCanMailboxCSCode::RxOverrun) => {
+                let id =
+                    unsafe { core::ptr::read_volatile((mailbox_addr + 0x4_u32) as *const u32) };
+                let data0 =
+                    unsafe { core::ptr::read_volatile((mailbox_addr + 0x8_u32) as *const u32) };
+                let data1 =
+                    unsafe { core::ptr::read_volatile((mailbox_addr + 0xC_u32) as *const u32) };
+
+                let mut data: [u8; 8] = [0, 0, 0, 0, 0, 0, 0, 0];
+                for i in 0..4 {
+                    data[3 - i] = (data0 >> (8 * i)) as u8;
+                }
+                for i in 0..4 {
+                    data[7 - i] = (data1 >> (8 * i)) as u8;
+                }
+
+                self.write_mailbox(
+                    mailbox_number,
+                    Some(FlexCanMailboxCSCode::RxEmpty.to_code_reg()),
+                    None,
+                    None,
+                );
+                read_reg!(ral::can, self.reg, TIMER);
+                self.write_iflag_bit(mailbox_number);
+
+                let frame = Frame::new_from_raw(code, id, data);
+
+                Some(MailboxData {
+                    frame,
+                    mailbox_number,
+                })
+            }
+            _ => None,
+        }
+    }
+
+    /// Write the registers of a mailbox.
+    #[inline(always)]
+    fn write_mailbox(
```

Comment:
There are FlexCAN errata, at least on the RT1064, that require some extra precautions and/or workarounds. This function and elsewhere in the module would be affected. Those are 5641 (not on RT1064), 5829, 6032, and 9595. Should this driver be expected to know of and handle these errata? Or would there be more board-specific drivers for that? Overall excellent work, just curious how something like that is handled.

Reply:
Thanks for helping with the review. If an erratum's workaround can act as the driver's default behavior, that would be ideal, and we should handle it here. This cuts down on configurations / code paths we maintain. But, if we need separate code paths to handle a driver's errata, we could consider crate configurations / separate driver crates / common- and chip-specific HAL combinations.